### PR TITLE
ddl, parser: support 'ALTER TABLE RENAME KEY TO' syntax

### PR DESCRIPTION
**Pull request by spongedu**
for #6405

### CODE REVIEW

**Comment by shenli** [comment]
@spongedu Thanks!
Please resolve the conflicts.

File: ddl/table.go [hunk 1]
```diff
@@ -356,6 +356,31 @@ func (d *ddl) onModifyTableComment(t *meta.Meta, job *model.Job) (ver int64, _ e
...
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	tblInfo, err := getTableInfo(t, job, job.SchemaID)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	idx := findIndexByName(from.L, tblInfo.Indices)
```
Comment (tiancaiamao) [inline ddl/table.go]: I'm afraid we have to double check here if we don't take @coocood 's  suggestion.
Take this scenario for example:
TiDB client1,  rename idx1 to idx3,   check success,  put into queue
TiDB client2,  rename idx2 to idx3,   check success,  put into queue

When the code runs to here, those two request would both success! 
We lost a index, no error!

File: parser/parser.y [hunk 2]
```diff
@@ -1036,6 +1036,14 @@ AlterTableSpec:
 			NewTable:      $3.(*ast.TableName),
 		}
 	}
+|	"RENAME" KeyOrIndex Identifier "TO" Identifier
+	{
+		$$ = &ast.AlterTableSpec{
+			Tp:    		ast.AlterTableRenameIndex,
```
[the remark this answers is not in the document]
Comment (spongedu) [reply parser/parser.y]: ok

File: ddl/table.go (hunk 1 above)
Comment (spongedu) [reply ddl/table.go]: yes, double check is needed here.

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -1559,6 +1561,42 @@ func (d *ddl) AlterTableComment(ctx sessionctx.Context, ident ast.Ident, spec *a
 	return errors.Trace(err)
 }
 
+// RenameIndex updates the table comment information.
```
Comment (lamxTyler) [inline ddl/ddl_api.go]: Is the comment not correct?

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -1559,6 +1561,42 @@ func (d *ddl) AlterTableComment(ctx sessionctx.Context, ident ast.Ident, spec *a
...
+		return errors.Trace(infoschema.ErrTableNotExists.GenByArgs(ident.Schema, ident.Name))
+	}
+
+	if fromIdx := findIndexByName(spec.FromKey.L, tb.Meta().Indices); fromIdx == nil {
+		return errors.Trace(infoschema.ErrKeyNotExists.GenByArgs(spec.FromKey.O, tb.Meta().Name))
+	}
+	if spec.FromKey.O == spec.ToKey.O {
+		return nil
+	}
+	if toIdx := findIndexByName(spec.ToKey.L, tb.Meta().Indices); toIdx != nil && spec.FromKey.L != spec.ToKey.L {
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply ddl/ddl_api.go]: So, add comment to the code.

[the remark this answers is not in the document]
Comment (spongedu) [reply ddl/ddl_api.go]: ok

**Comment by tiancaiamao** [comment]
LGTM

File: ddl/table.go [hunk 5]
```diff
@@ -357,6 +357,37 @@ func (d *ddl) onModifyTableComment(t *meta.Meta, job *model.Job) (ver int64, _ e
...
+	}
+	tblInfo, err := getTableInfo(t, job, job.SchemaID)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	// Double check. See function `RenameIndex` in ddl_api.go
+	idx := findIndexByName(from.L, tblInfo.Indices)
+	if idx == nil {
+		return ver, errors.Trace(infoschema.ErrKeyNotExists.GenByArgs(from, tblInfo.Name))
```
Comment (ciscoxll) [inline ddl/table.go]: Ditto.

File: ddl/table.go [hunk 6]
```diff
@@ -357,6 +357,37 @@ func (d *ddl) onModifyTableComment(t *meta.Meta, job *model.Job) (ver int64, _ e
...
+	if i := findIndexByName(to.L, tblInfo.Indices); i != nil && from.L != to.L {
+		return ver, errors.Trace(infoschema.ErrKeyNameDuplicate.GenByArgs(to.O))
+	}
+	idx.Name = to
+	ver, err = updateVersionAndTableInfo(t, job, tblInfo, true)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	job.FinishTableJob(model.JobStateDone, model.StatePublic, ver, tblInfo)
+	return ver, nil
```
[the remark this answers is not in the document]
Comment (winkyao) [reply ddl/table.go]: I think return nil is ok.

**Comment by shenli** [comment]
@zimulala PTAL

File: ddl/table.go [hunk 7]
```diff
@@ -357,6 +357,37 @@ func (d *ddl) onModifyTableComment(t *meta.Meta, job *model.Job) (ver int64, _ e
...
 
+func (d *ddl) onRenameIndex(t *meta.Meta, job *model.Job) (ver int64, _ error) {
+	var from, to model.CIStr
+	if err := job.DecodeArgs(&from, &to); err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+	tblInfo, err := getTableInfo(t, job, job.SchemaID)
+	if err != nil {
+		return ver, errors.Trace(err)
```
[the remark this answers is not in the document]
Comment (spongedu) [reply ddl/table.go]: OK

File: ddl/table.go (hunk 5 above)
Comment (spongedu) [reply ddl/table.go]: OK

**Comment by spongedu** [comment]
@ciscoxll  PTAL

**Comment by spongedu** [comment]
@lamxTyler  PTAL, thx

File: parser/parser.y [hunk 8]
```diff
@@ -1040,6 +1040,14 @@ AlterTableSpec:
...
 		}
 	}
+|	"RENAME" KeyOrIndex Identifier "TO" Identifier
+	{
+		$$ = &ast.AlterTableSpec{
+			Tp:    	    ast.AlterTableRenameIndex,
+			FromKey:    model.NewCIStr($3),
+			ToKey:      model.NewCIStr($5),
+		}
+	}
```
Comment (zimulala) [inline parser/parser.y]: Please add tests in parser_test.go

File: executor/executor_test.go [hunk 9]
```diff
@@ -2645,6 +2645,51 @@ func (s *testSuite) TestCheckIndex(c *C) {
...
+	// Test rename to the same name
+	_, err = se.Execute(context.Background(), "alter table t rename index k3 to k3")
+	c.Assert(err, IsNil)
+	_, err = se.Execute(context.Background(), "admin check index t k3")
+	c.Assert(err, IsNil)
+
+	// Test rename on non-exists keys
+	_, err = se.Execute(context.Background(), "alter table t rename index x to x")
+	c.Assert(err, NotNil)
+	c.Assert(strings.Contains(err.Error(), "Key 'x' doesn't exist in table 't'"), IsTrue)
```
Comment (zimulala) [inline executor/executor_test.go]: You can use `testErrorCode` in this file.

File: ddl/table.go [hunk 10]
```diff
@@ -357,6 +357,40 @@ func (d *ddl) onModifyTableComment(t *meta.Meta, job *model.Job) (ver int64, _ e
...
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+	tblInfo, err := getTableInfo(t, job, job.SchemaID)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	// Double check. See function `RenameIndex` in ddl_api.go
```
Comment (zimulala) [inline ddl/table.go]: Could you extract this code into a function?

File: parser/parser.y (hunk 8 above)
Comment (spongedu) [reply parser/parser.y]: ok

File: executor/executor_test.go (hunk 9 above)
Comment (spongedu) [reply executor/executor_test.go]: I'll move these test to `ddl/ddl_db_test.go`

File: ddl/table.go (hunk 10 above)
Comment (spongedu) [reply ddl/table.go]: ok

File: ddl/table.go [hunk 11]
```diff
@@ -357,6 +357,37 @@ func (d *ddl) onModifyTableComment(t *meta.Meta, job *model.Job) (ver int64, _ e
...
+	duplicate, err := d.validateRenameIndex(from, to, tblInfo)
+	if duplicate {
+		return ver, nil
+	}
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, err
```
Comment (zimulala) [inline ddl/table.go]: s/err/error.Trace(err)

File: ddl/ddl_api.go [hunk 12]
```diff
@@ -1574,6 +1576,58 @@ func (d *ddl) AlterTableComment(ctx sessionctx.Context, ident ast.Ident, spec *a
...
+	tb, err := is.TableByName(ident.Schema, ident.Name)
+	if err != nil {
+		return errors.Trace(infoschema.ErrTableNotExists.GenByArgs(ident.Schema, ident.Name))
+	}
+	duplicate, err := d.validateRenameIndex(spec.FromKey, spec.ToKey, tb.Meta())
+	if duplicate {
+		return nil
+	}
+	if err != nil {
+		return err
```
Comment (zimulala) [inline ddl/ddl_api.go]: s/err/error.Trace(err)

File: ddl/ddl_api.go [hunk 13]
```diff
@@ -1574,6 +1576,58 @@ func (d *ddl) AlterTableComment(ctx sessionctx.Context, ident ast.Ident, spec *a
...
+		BinlogInfo: &model.HistoryInfo{},
+		Args:       []interface{}{spec.FromKey, spec.ToKey},
+	}
+
+	err = d.doDDLJob(ctx, job)
+	err = d.callHookOnChanged(err)
+	return errors.Trace(err)
+}
+
+func (d *ddl) validateRenameIndex(from, to model.CIStr, tbl *model.TableInfo) (duplicate bool, err error) {
```
Comment (zimulala) [inline ddl/ddl_api.go]: Could you put this function to the file of index.go?

Comment (spongedu) [reply ddl/ddl_api.go]: ok

**Comment by spongedu** [comment]
Done @zimulala  PTAL :)

File: ddl/index.go [hunk 14]
```diff
@@ -175,6 +176,23 @@ func dropIndexColumnFlag(tblInfo *model.TableInfo, indexInfo *model.IndexInfo) {
 	}
 }
 
+func validateRenameIndex(from, to model.CIStr, tbl *model.TableInfo) (duplicate bool, err error) {
```
Comment (shenli) [inline ddl/index.go]: `duplicate` is not a proper name for the first return value. It confused me. I thought it is the case that the to.L is the same as any other index name.

Comment (shenli) [reply ddl/index.go]: `ignore` is better

Comment (spongedu) [reply ddl/index.go]: reasonable. I've fixed this. :)

**Comment by shenli** [comment]
LGTM

**Pull request merged**